plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/CabalAdd.hs

```diff
@@ -0,0 +1,185 @@
+        command = mkLspCommand plId (CommandId cabalAddCommand) "Execute Code Action" (Just [toJSON params])
+      in CodeAction title (Just CodeActionKind_QuickFix) (Just []) Nothing Nothing Nothing (Just command) Nothing
+
+-- | Gives a mentioned number of hidden packages given
+--   a specific error message
+hiddenPackageSuggestion :: Int -> T.Text -> [(T.Text, T.Text)]
+hiddenPackageSuggestion maxCompletions msg = take maxCompletions $ getMatch (msg =~ regex)
+  where
+    regex :: T.Text -- TODO: Support multiple packages suggestion
+    regex = "Could not load module \8216.*\8217.\nIt is a member of the hidden package [\8216\\']([a-z]+)[-]?([0-9\\.]*)[\8217\\']"
+    getMatch :: (T.Text, T.Text, T.Text, [T.Text]) -> [(T.Text, T.Text)]
+    getMatch (_, _, _, []) = []
+    getMatch (_, _, _, [dependency]) = [(dependency, T.empty)]
+    getMatch (_, _, _, [dependency, version]) = [(dependency, version)]
+    getMatch (_, _, _, _) = error "Impossible pattern matching case"
+
+
+cabalAddCommand :: IsString p => p
+cabalAddCommand = "cabalAdd"
+
+data CabalAddCommandParams =
+     CabalAddCommandParams { cabalPath   :: FilePath
+                           , buildTarget :: Maybe String
+                           , dependency  :: T.Text
+                           , version     :: Maybe T.Text
+                           }
+  deriving (Generic, Show)
+  deriving anyclass (FromJSON, ToJSON)
+
+command :: CommandFunction IdeState CabalAddCommandParams
+command _ _ (CabalAddCommandParams {cabalPath = path, buildTarget = target, dependency = dep, version = mbVer}) = do
+  let specifiedDep = case mbVer of
+        Nothing  -> dep
+        Just ver -> dep <> " ^>=" <> ver
+  void $ liftIO $ addDependency path target (fromList [T.unpack specifiedDep])
+  pure $ InR Null
+
+-- | Gives cabal file's contents or throws error.
+--   Inspired by @readCabalFile@ in cabal-add,
+--   Distribution.Client.Main
+readCabalFile :: FilePath -> IO ByteString
+readCabalFile fileName = do
+  cabalFileExists <- doesFileExist fileName
+  if cabalFileExists
+    then snd . patchQuirks <$> B.readFile fileName
+    else error ("Failed to read cabal file at " <> fileName)
+
+getBuildTargets :: GenericPackageDescription -> FilePath -> FilePath -> IO [BuildTarget]
+getBuildTargets gpd cabalFilePath haskellFilePath = do
+  let haskellFileRelativePath = makeRelative (dropFileName cabalFilePath) haskellFilePath
+  readBuildTargets (verboseNoStderr silent) (flattenPackageDescription gpd) [haskellFileRelativePath]
+
+
+-- | Constructs prerequisets for the @executeConfig@
+--   and runs it, given path to the cabal file and a dependency message.
+--
+--   Inspired by @main@ in cabal-add,
+--   Distribution.Client.Main
+addDependency :: FilePath -> Maybe String -> NonEmpty String -> IO ()
+addDependency cabalFilePath buildTarget dependency = do
+
+  cnfOrigContents <- readCabalFile cabalFilePath
```

Comment:
I don't know if HLS does it in other cases, but it could be a good idea to lock the handle using https://hackage.haskell.org/package/lukko-0.1.2/docs/Lukko.html#v:hLock.

Reply:
Or without `lukko`, open file for read/write access and manipulate it without closing?.. Or am I talking nonsense?..

Reply:
Instead of manually reading and parsing, use https://github.com/haskell/haskell-language-server/blob/master/plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs#L260, which hooks into the rule system.

E.g. 
```haskell
liftIO $ runAction "cabal-plugin.cabal-add" ide $ useWithStale ParseCabalFile $ toNormalizedFilePath $ toNormalizedFilePath cabalFilePath
```
This will give you the GPD already parsed and cached.

Reply:
Or if you need the fields, the original link will give you that

Reply:
Since the `handle` using approach is not very in the style of hls I used `runAction` approach instead. It falls back on reading files, if the data wasn't found in rules.